utils/filesystem/files.go

```diff
@@ -1129,6 +1128,60 @@ func (fs *VFS) LsWithExclusionPatterns(dir string, exclusionPatterns ...string)
 	return
 }
 
+func (fs *VFS) LsRecursive(ctx context.Context, dir string) (files []string, err error) {
+	limits := NoLimits()
+	return fs.LsRecursiveWithExtensionPatternsAndLimits(ctx, dir, limits)
+}
+
+func (fs *VFS) LsRecursiveWithExtensionPatterns(ctx context.Context, dir string, exclusionPatterns ...string) (files []string, err error) {
+	limits := NoLimits()
+	return fs.LsRecursiveWithExtensionPatternsAndLimits(ctx, dir, limits, exclusionPatterns...)
+}
+
+func (fs *VFS) LsRecursiveWithExtensionPatternsAndLimits(ctx context.Context, dir string, limits ILimits, exclusionPatterns ...string) (files []string, err error) {
+	err = parallelisation.DetermineContextError(ctx)
+	if err != nil {
+		return
+	}
+	if limits == nil {
+		err = fmt.Errorf("%w: missing file system limits definition", commonerrors.ErrUndefined)
+		return
+	}
+	exclusionRegex, err := NewExclusionRegexList(fs.PathSeparator(), exclusionPatterns...)
+	if err != nil {
+		return
+	}
+	if IsPathExcluded(dir, exclusionRegex...) {
+		// In this case, the whole tree is excluded and so, we can stop here
+		return
+	}
+
+	fn := func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		relativePath := strings.TrimPrefix(path, dir)
+		currentDepth := int64(strings.Count(relativePath, string(fs.PathSeparator())))
```

Comment:
this might not work if you escape the path separator, e.g.:


`/dir1/dir2` == depth 2
`/dir1\/escaped` == depth 1 but by just counting the separators you will have depth 2 which is wrong

Reply:
You should turn it into a path object instead of doing string operations